### PR TITLE
Descent and Hybrid Optimization for Excited States

### PR DESCRIPTION
**Pull request by LeonOtis**
## Proposed changes

These changes extend the accelerated descent and hybrid optimizers introduced in #1750 for ground state energy minimization to minimizing the excited state functional. Most of the changes in the code relate to the DescentEngine object. Details of the descent and hybrid methods are in a recent paper: https://arxiv.org/abs/2008.03586 
I'm happy to clarify anything about the methods themselves or the code and really appreciate any feedback. 

## What type(s) of changes does this code introduce?

- New feature

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?

Savio computing cluster

## Checklist

- Yes. This PR is up to date with the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by prckent** [comment]
OK to test

**Comment by prckent** [comment]
Thanks Leon. Does this change any of the ground state behaviors at all?

**Comment by LeonOtis** [comment]
It should not affect the ground state optimization. Optimizing the excited state functional with descent/hybrid has to be switched on by user input in the same way as with the linear method.

**Review by prckent — COMMENTED** [review]
I tidied the bibliography a little, putting a link to your preprint.

**Review by prckent — CHANGES_REQUESTED** [review]
Since it looks like this doesn't work with complex, please can you add a sanity check abort, unless there is one already.

**Comment by ye-luo** [comment]
Test this please.

**Review by prckent — DISMISSED** [review]

**Comment by prckent** [comment]
(we'll wait for Ye's review)

**Comment by prckent** [comment]
@ye-luo OK?

**Comment by ye-luo** [comment]
Review will happen after lunch.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimize.cpp
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimize.cpp:205]: The communicator is accessible in this class. Use `myComm->barrier_and_abort()` to have clean print.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimize.cpp
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimize.cpp:213]: Same error printing here.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimize.cpp
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimize.cpp:511]: Same error printing here.

File: src/QMCDrivers/Optimizers/DescentEngine.h
Comment (ye-luo) [inline src/QMCDrivers/Optimizers/DescentEngine.h:70]: `//` is for generic source code commenting and omitted by by doxygen. use `///` to document variables and functions. Please update throughout this file.

File: src/QMCDrivers/Optimizers/DescentEngine.h
Comment (ye-luo) [inline src/QMCDrivers/Optimizers/DescentEngine.h:254]: please expand existing unit test to cover this function.

**Review by ye-luo — CHANGES_REQUESTED** [review]
This PR is closed to be merged. Unit test needs to be expanded.

Is it difficult to add a integration test? It is still desired.
PS: I noticed that short-H4-hybrid seems broken (probably by me).
short-H4-descent works fine.

**Comment by LeonOtis** [comment]
Thanks for the review Ye. I've made the changes and also added integration tests short-H4-target-descent and short-H4-target-hybrid. For my build, trying the integration tests with ctest hits a problem with a mismatch in the hdf5 version ("Headers are 1.10.1, library is 1.10.5") and aborts. I think 1.10.5 is the only version I have loaded and setting HDF5_DISABLE_VERSION_CHECK=1 seems to cause the test to hang. However, if I perform the test by hand (run qmcpack with the test input and then use the check_scalars.py script) things seem to work and properly pass both for the new tests and the short-H4-descent and short-H4-hybrid.

File: src/QMCDrivers/Optimizers/DescentEngine.cpp
Comment (ye-luo) [inline src/QMCDrivers/Optimizers/DescentEngine.cpp:427]: Without the protection, does the code run into problems? It is supposed to work unless my_comm was not set up properly.

File: src/QMCDrivers/tests/test_DescentEngine.cpp
Comment (ye-luo) [inline src/QMCDrivers/tests/test_DescentEngine.cpp:31]: Change to `Communicate* c = OHMMS::Controller;` and pass c to the constructor. This may solve your MPI issue in unit test.

**Comment by ye-luo** [comment]
> Thanks for the review Ye. I've made the changes and also added integration tests short-H4-target-descent and short-H4-target-hybrid. For my build, trying the integration tests with ctest hits a problem with a mismatch in the hdf5 version ("Headers are 1.10.1, library is 1.10.5") and aborts. I think 1.10.5 is the only version I have loaded and setting HDF5_DISABLE_VERSION_CHECK=1 seems to cause the test to hang. However, if I perform the test by hand (run qmcpack with the test input and then use the check_scalars.py script) things seem to work and properly pass both for the new tests and the short-H4-descent and short-H4-hybrid.

This happens when you reuse a build folder. The hdf5 header file path was recorded from your old build while your modules are updated to a newer version. Simply rebuilding from scratch should resolve this issue.

File: src/QMCDrivers/tests/test_DescentEngine.cpp
Comment (LeonOtis) [reply src/QMCDrivers/tests/test_DescentEngine.cpp]: Thanks Ye. This fixes the MPI issue in the unit test so I've removed the check in DescentEngine.

**Review by ye-luo — APPROVED** [review]
The data center hosting CI is under power work. Will merge once it is back.

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
@LeonOtis When I was looking at the test failing record, I noticed this PR breaks
https://cdash.qmcpack.org/CDash/testDetails.php?test=12217349&build=191149
https://cdash.qmcpack.org/CDash/testDetails.php?test=12217353&build=191149

**Comment by LeonOtis** [comment]
@ye-luo Thanks for letting me know, Ye. Is this related to the bugs you fixed in #3031 ? The tests with descent and the hybrid method are with a MSD wave function and they seem to work properly for me when I use the latest version of develop with those fixes. If they are still breaking for you, let me know.

**Comment by ye-luo** [comment]
@LeonOtis #3031 fixed a bug introduced this month but the 1-16 case failure started with this PR merged in. From last night, 16-1 is good https://cdash.qmcpack.org/CDash/testDetails.php?test=12517173&build=196342
but not 1-16 https://cdash.qmcpack.org/CDash/testDetails.php?test=12517174&build=196342

**Pull request merged**
